In [1]:
import findspark
findspark.init()

import pyspark
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

23/03/08 21:05:16 WARN Utils: Your hostname, Mac-Pro-Olga.local resolves to a loopback address: 127.0.0.1; using 192.168.0.161 instead (on interface en0)
23/03/08 21:05:16 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address


Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


23/03/08 21:05:17 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
23/03/08 21:05:18 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.


Question 1:
Install Spark and PySpark

Install Spark
Run PySpark
Create a local spark session
Execute spark.version.
What's the output?

A 1: 3.3.2

In [2]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz


--2023-03-08 21:05:25--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/fhvhv/fhvhv_tripdata_2021-06.csv.gz
Распознаётся github.com (github.com)… 140.82.121.4
Подключение к github.com (github.com)|140.82.121.4|:443... соединение установлено.
HTTP-запрос отправлен. Ожидание ответа… 302 Found
Адрес: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/4564ad9e-a6da-4923-ad6f-35ff02446a51?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20230308%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230308T180526Z&X-Amz-Expires=300&X-Amz-Signature=554acddcb2494be209606e98ef8475d0bb6c025d47634f08069a95b85a5a5e98&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dfhvhv_tripdata_2021-06.csv.gz&response-content-type=application%2Foctet-stream [переход]
--2023-03-08 21:05:26--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948

In [5]:
!wc -l fhvhv_tripdata_2021-06.csv.gz


  651315 fhvhv_tripdata_2021-06.csv.gz


In [38]:
df = spark.read \
    .option("header", "true") \
    .csv('fhvhv_tripdata_2021-06.csv')

df.schema

StructType([StructField('dispatching_base_num', StringType(), True), StructField('pickup_datetime', StringType(), True), StructField('dropoff_datetime', StringType(), True), StructField('PULocationID', StringType(), True), StructField('DOLocationID', StringType(), True), StructField('SR_Flag', StringType(), True), StructField('Affiliated_base_number', StringType(), True)])

In [53]:
from pyspark.sql import types
schema = types.StructType([
    #types.StructField('hvfhs_license_num', types.StringType(), True),
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropoff_datetime', types.TimestampType(), True),
    types.StructField('PULocationID', types.IntegerType(), True),
    types.StructField('DOLocationID', types.IntegerType(), True),
    types.StructField('SR_Flag', types.StringType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True)
])
df = spark.read \
    .option("header", "true") \
    .schema(schema) \
    .csv('fhvhv_tripdata_2021-06.csv')
df = df.repartition(12)
df.write.parquet('fhvhv/2021/06/')

23/03/08 21:40:55 WARN MemoryManager: Total allocation exceeds 95,00% (1 020 054 720 bytes) of heap memory
Scaling row group sizes to 95,00% for 8 writers
23/03/08 21:40:55 WARN MemoryManager: Total allocation exceeds 95,00% (1 020 054 720 bytes) of heap memory
Scaling row group sizes to 84,44% for 9 writers
23/03/08 21:40:55 WARN MemoryManager: Total allocation exceeds 95,00% (1 020 054 720 bytes) of heap memory
Scaling row group sizes to 76,00% for 10 writers


23/03/08 21:40:57 WARN MemoryManager: Total allocation exceeds 95,00% (1 020 054 720 bytes) of heap memory
Scaling row group sizes to 84,44% for 9 writers
23/03/08 21:40:57 WARN MemoryManager: Total allocation exceeds 95,00% (1 020 054 720 bytes) of heap memory
Scaling row group sizes to 95,00% for 8 writers


Question 2:
HVFHW June 2021

Read it with Spark using the same schema as we did in the lessons.
We will use this dataset for all the remaining questions.
Repartition it to 12 partitions and save it to parquet.
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

A 2: 24MB


In [54]:
df2 = spark.read.parquet('fhvhv/2021/06/')


In [55]:
from pyspark.sql import functions as F
df2.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B02510|2021-06-03 21:49:07|2021-06-03 22:02:01|         262|          79|      N|                  null|
|              B02878|2021-06-02 17:18:16|2021-06-02 17:30:23|         116|          24|      N|                B02878|
|              B02510|2021-06-02 06:15:09|2021-06-02 06:20:26|          97|          65|      N|                  null|
|              B02510|2021-06-01 17:36:28|2021-06-01 17:50:42|         230|          90|      N|                  null|
|              B02510|2021-06-01 16:58:18|2021-06-01 17:05:29|          79|         148|      N|                  null|
|              B02765|2021-06-02 00:17:4

Question 3:
Count records

How many taxi trips were there on June 15?

Consider only trips that started on June 15.

A3: 452,470

In [56]:
from datetime import datetime 

In [57]:
df2.filter((df2.pickup_datetime >= datetime(2021, 6, 15))&(df2.pickup_datetime < datetime(2021, 6, 16))).count()

452470

Question 4:
Longest trip for each day

Now calculate the duration for each trip.
How long was the longest trip in Hours?

A 4: 66.87 Hours

In [89]:
df2.select(
      (F.max(F.col("dropoff_datetime").cast("long")-F.col("pickup_datetime").cast("long"))/3600).alias("datediff")
    ).show()

+----------------+
|        datediff|
+----------------+
|66.8788888888889|
+----------------+



Question 5:
User Interface

Spark’s User Interface which shows application's dashboard runs on which local port?

A 5: 4040

Question 6:
Most frequent pickup location zone

Load the zone lookup data into a temp view in Spark
Zone Data

Using the zone lookup data and the fhvhv June 2021 data, what is the name of the most frequent pickup location zone?

A 6: Crown Heights North

In [90]:
df_zones = spark.read.parquet('zones/')


In [93]:
df_result = df2.join(df_zones, df2.PULocationID == df_zones.LocationID)


In [95]:
df_result.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+----------+-------------+--------------------+------------+
|dispatching_base_num|    pickup_datetime|   dropoff_datetime|PULocationID|DOLocationID|SR_Flag|Affiliated_base_number|LocationID|      Borough|                Zone|service_zone|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+----------+-------------+--------------------+------------+
|              B02510|2021-06-03 21:49:07|2021-06-03 22:02:01|         262|          79|      N|                  null|       262|    Manhattan|      Yorkville East| Yellow Zone|
|              B02878|2021-06-02 17:18:16|2021-06-02 17:30:23|         116|          24|      N|                B02878|       116|    Manhattan|    Hamilton Heights|   Boro Zone|
|              B02510|2021-06-02 06:15:09|2021-06-02 06:20:26|          97|          65|      N|         

In [94]:
df_result.registerTempTable('trips')


In [100]:
df_revenue = spark.sql("""
SELECT     Zone AS zone,
    COUNT(1) AS number_records
FROM
    trips
GROUP BY
    1
order by 2 desc
""")

In [101]:
df_revenue.show()

+--------------------+--------------+
|                zone|number_records|
+--------------------+--------------+
| Crown Heights North|        231279|
|        East Village|        221244|
|         JFK Airport|        188867|
|      Bushwick South|        187929|
|       East New York|        186780|
|TriBeCa/Civic Center|        164344|
|   LaGuardia Airport|        161596|
|            Union Sq|        158937|
|        West Village|        154698|
|             Astoria|        152493|
|     Lower East Side|        151020|
|        East Chelsea|        147673|
|Central Harlem North|        146402|
|Williamsburg (Nor...|        143683|
|          Park Slope|        143594|
|  Stuyvesant Heights|        141427|
|        Clinton East|        139611|
|West Chelsea/Huds...|        139431|
|             Bedford|        138428|
|         Murray Hill|        137879|
+--------------------+--------------+
only showing top 20 rows

